### PR TITLE
test: add test to prove #596 is fixed

### PR DESCRIPTION
**Pull request by big-andy-coates**
### Description 

fixes: https://github.com/confluentinc/ksql/issues/596

Functionality to join on an expression was added previously.

### Testing done 

Test only change.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksql-functional-tests/src/test/resources/query-validation-tests/joins.json [hunk 1]
```diff
@@ -1572,11 +1572,26 @@
...
+        {"topic": "left_topic", "value": {"id": 1}, "timestamp": 10},
+        {"topic": "right_topic", "value": {"id": 1}, "timestamp": 10}
       ],
       "outputs": [
-        {"topic": "OUTPUT", "key": 1, "value": {"T_ID": 1}, "timestamp": 10}
+        {"topic": "OUTPUT", "key": 1, "value": {"T_ID": 1}, "timestamp": 11}
+      ]
+    },
+    {
+      "name": "stream stream join - contains CAST double to int",
```
Comment (agavra) [inline ksql-functional-tests/src/test/resources/query-validation-tests/joins.json]: we already have a test for this:
```
    {
      "name": "stream stream join - contains CAST",
      "statements": [
        "CREATE STREAM TEST1 (ID bigint) WITH (kafka_topic='left_topic', value_format='JSON');",
        "CREATE STREAM TEST2 (ID int) WITH (kafka_topic='right_topic', value_format='JSON');",
        "CREATE STREAM OUTPUT as SELECT t.ID FROM test1 t JOIN test2 tt WITHIN 30 seconds ON t.id = CAST(tt.id AS BIGINT);"
      ],
      "inputs": [
        {"topic": "left_topic", "key": "1", "value": {"id": 1}, "timestamp": 10},
        {"topic": "right_topic", "key": "1", "value": {"id": 1}, "timestamp": 10}
      ],
      "outputs": [
        {"topic": "OUTPUT", "key": 1, "value": {"T_ID": 1}, "timestamp": 10}
      ]
    },
```

Comment (agavra) [reply ksql-functional-tests/src/test/resources/query-validation-tests/joins.json]: see https://github.com/confluentinc/ksql/pull/4278/files#diff-27daef0b162db054c2579c7937b76156 for all tests added

Comment (big-andy-coates) [reply ksql-functional-tests/src/test/resources/query-validation-tests/joins.json]: I know - but an int -> long cast is fairly trivial: the value doesn't change.  Wanted to just ensure the value actually changed.

Comment (agavra) [reply ksql-functional-tests/src/test/resources/query-validation-tests/joins.json]: fair enough!

**Pull request merged**